tsl/src/nodes/data_node_copy.c

```diff
@@ -0,0 +1,414 @@
+				{
+					ExprContext *econtext;
+
+					Assert(NULL != rri_saved->ri_projectReturning);
+					econtext = rri_saved->ri_projectReturning->pi_exprContext;
+					econtext->ecxt_scantuple = slot;
+				}
+
+				if (dncs->set_processed)
+					estate->es_processed++;
+			}
+		}
+	} while (!has_returning && !TupIsNull(slot));
+
+	/* Reset the result relation to point to the root hypertable before
+	 * returning, since the child ChunkDispatch node set it to the chunk. */
+	estate->es_result_relation_info = rri_saved;
+
+	/* Tuple routing in the ChunkDispatchState subnode sets the result
+	 * relation to a chunk when routing, but the read handler should have
+	 * ensured the result relation is reset. */
+	Assert(node->ss.ps.state->es_result_relation_info->ri_RelationDesc->rd_id == dncs->rel->rd_id);
+	Assert(node->ss.ps.state->es_result_relation_info->ri_usesFdwDirectModify);
+
+	return slot;
+}
+
+static void
+data_node_copy_rescan(CustomScanState *node)
+{
+	/* Cannot rescan and start from the beginning since we might already have
+	 * sent data to remote nodes */
+	elog(ERROR, "cannot restart inserts to remote nodes");
+}
+
+static void
+data_node_copy_end(CustomScanState *node)
+{
+	DataNodeCopyState *dncs = (DataNodeCopyState *) node;
+
+	ExecEndNode(linitial(node->custom_ps));
+	remote_copy_end(dncs->copy_ctx);
+	ts_cache_release(dncs->hcache);
+}
+
+static void
+data_node_copy_explain(CustomScanState *node, List *ancestors, ExplainState *es)
+{
+	DataNodeCopyState *dncs = (DataNodeCopyState *) node;
+	/*
+	 * Add remote COPY command when VERBOSE option is specified.
+	 */
+	if (es->verbose)
+		ExplainPropertyText("Remote SQL", remote_copy_get_copycmd(dncs->copy_ctx), es);
+}
+
+static CustomExecMethods data_node_copy_state_methods = {
+	.CustomName = "DataNodeCopyState",
+	.BeginCustomScan = data_node_copy_begin,
+	.EndCustomScan = data_node_copy_end,
+	.ExecCustomScan = data_node_copy_exec,
+	.ReScanCustomScan = data_node_copy_rescan,
+	.ExplainCustomScan = data_node_copy_explain,
+};
+
+/* Only allocate the custom scan state. Initialize in the begin handler. */
+static Node *
+data_node_copy_state_create(CustomScan *cscan)
+{
+	DataNodeCopyState *dncs;
+
+	dncs = (DataNodeCopyState *) newNode(sizeof(DataNodeCopyState), T_CustomScanState);
+	dncs->cstate.methods = &data_node_copy_state_methods;
+
+	return (Node *) dncs;
+}
+
+static CustomScanMethods data_node_copy_plan_methods = {
+	.CustomName = "DataNodeCopy",
+	.CreateCustomScanState = data_node_copy_state_create,
+};
+
+/*
+ * Check for a user-created (custom) type.
+ *
+ * This is an imperfect check; we treat any composite type as
+ * user-created. Howevever, there are user-created types that aren't composite
+ * types (e.g., those implemented in C). There's currently no way to reliably
+ * separate between native PostgreSQL types with "fixed" Oids,
+ * vs. user-created types with generated Oids.
+ */
+static bool type_is_custom(Oid typeid) { return type_is_rowtype(typeid); }
```

Comment:
I had suggested `type_is_rowtype(typeid) && typeid > MyDatabaseId`. Doesn't this correctly ensure that the type is a user created custom datatype? Anyways, it's not a big deal as discussed earlier.

Reply:
I don't think so. rowtypes only cover composite types and not other custom types (e.g., varlen types implemented in C). 

I didn't understand the typeid > MyDatabaseId comparison. Why would it make sense to compare a type OID with a Database OID?

Admittedly, this type checking is imperfect, but I don't think there's a consistent/reliable way to know that a type is user-created.

Reply:
The database has to be created FIRST with a lower OID, any new objects in this database will have OID more than the dboid was my point.

Reply:
Plus for this use of `FirstNormalObjectId`